plutarch-test/src/Plutarch/Test/Property/Util.hs

```diff
@@ -0,0 +1,286 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Test.Property.Util (
+  Marshal (marshal),
+  NotLambda,
+  haskPlutEquiv,
+  leftInverse,
+  viaData,
+  viaPEq,
+  viaBoth,
+  viaPEqPartial,
+  viaDataPartial,
+  viaBothPartial,
+  testDataEq,
+  testPEq,
+  run,
+) where
+
+import Plutarch.Prelude
+
+import Plutarch (ClosedTerm, compile)
+
+import Plutarch.Evaluate (evaluateScript)
+
+import Control.Exception (SomeException, evaluate, try)
+import Control.Monad.IO.Class (liftIO)
+import Data.Proxy (Proxy (..))
+import Data.Text (Text)
+
+import Plutus.V1.Ledger.Scripts (Script (..), ScriptError)
+import PlutusCore.Evaluation.Machine.ExBudget (ExBudget)
+
+import Hedgehog (Gen, Property, PropertyT, annotate, annotateShow, assert, forAll, property, (===))
+
+class Marshal h (p :: PType) | h -> p where
```

Comment:
How is this class related to this PR?

Reply:
When testing agreement between Haskell functions and Plutarch functions it's convenient to be able to generate the arguments to the function as Haskell terms and marshal them into Plutarch terms, that way you have both representations and can run both functions. Marshal is also used to turn the output of the Haskell function into Plutarch to test equality at the end.